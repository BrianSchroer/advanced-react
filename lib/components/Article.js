import React from 'react';
import PropTypes from 'prop-types';

const Article = ({article, author}) => {
  return (
    <div>
      <div>{article.title}</div>
      <div>{article.date}</div>
      <div>
        <a href={author.website} target="_blank">
          {author.firstName} {author.lastName}
        </a>
      </div>
      <div>{article.body}</div>
    </div>
  );
};

Article.propTypes = {
  article: PropTypes.object.isRequired,
  author: PropTypes.object.isRequired
};

export default Article;