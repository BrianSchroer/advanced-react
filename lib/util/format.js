export default class {
  static date(dateString) {
    return new Date(dateString).toDateString();
  }
}
