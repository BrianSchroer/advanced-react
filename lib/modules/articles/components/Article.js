import React from 'react';
import PropTypes from 'prop-types';
import glamorous from 'glamorous';
import format from '../../../util/format';

const StyledArticle = glamorous.div({
  paddingBottom: 10,
  borderBottomStyle: 'solid',
  borderBottomColor: '#aaa',
  borderBottomWidth: 1,
  marginBottom: 10
});

const ArticleTitle = glamorous.div({
  fontWeight: 'bold'
});

const ArticleDate = glamorous.div({
  fontSize: '0.85em',
  color: '#888'
});

const Author = glamorous.div({
  paddingTop: 10,
  paddingBottom: 10
});

const ArticleBody = glamorous.div({
  paddingLeft: 20
});

const Article = (props, context) => {
 
  const {article} = props;
  const author = context.store.lookUpAuthor(article.authorId);

  return (
    <StyledArticle>
      <ArticleTitle>{article.title}</ArticleTitle>
      <ArticleDate>{format.date(article.date)}</ArticleDate>
      <Author>
        <a href={author.website} target="_blank">
          {author.firstName} {author.lastName}
        </a>
      </Author>
      <ArticleBody>{article.body}</ArticleBody>
    </StyledArticle>
  );
};

Article.propTypes = {
  article: PropTypes.shape({
    date: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    body: PropTypes.string.isRequired
  })
};

Article.contextTypes = {
  store: PropTypes.object.isRequired
};

export default Article;