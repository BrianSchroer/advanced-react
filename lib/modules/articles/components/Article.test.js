import React from 'react';
import { shallow, renderer, snapshotHelper } from '../../../util/testHelpers';
import Article from './Article';

const testAuthor = {
  website: 'testWebsite',
  firstName: 'testFirstName',
  lastName: 'testLastName'
};

const testProps = {
  article: {
    title: 'testTitle',
    date: 'Mon Jul 11 2016 00:00:00 GMT+0000 (UTC)',
    body: 'testBody'
  },
};

const context = {
  store: {
    lookUpAuthor: jest.fn(()=> testAuthor)
  }
};

describe('Article', () => {

  it('renders correctly', () => {
    const wrapper = shallow(<Article {...testProps} />, { context });

    expect(wrapper).toMatchSnapshot();

    expect(context.store.lookUpAuthor).toHaveBeenCalledTimes(1);
  });

});