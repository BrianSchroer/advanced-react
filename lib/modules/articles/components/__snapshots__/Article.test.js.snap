// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <glamorous(div)>
    <glamorous(div)>
        testTitle
    </glamorous(div)>
    <glamorous(div)>
        Sun Jul 10 2016
    </glamorous(div)>
    <glamorous(div)>
        <a
            href="testWebsite"
            target="_blank"
        >
            testFirstName
             
            testLastName
        </a>
    </glamorous(div)>
    <glamorous(div)>
        testBody
    </glamorous(div)>
</glamorous(div)>,
  "nodes": Array [
    <glamorous(div)>
      <glamorous(div)>
            testTitle
      </glamorous(div)>
      <glamorous(div)>
            Sun Jul 10 2016
      </glamorous(div)>
      <glamorous(div)>
            <a
                  href="testWebsite"
                  target="_blank"
            >
                  testFirstName
                   
                  testLastName
            </a>
      </glamorous(div)>
      <glamorous(div)>
            testBody
      </glamorous(div)>
</glamorous(div)>,
  ],
  "options": Object {
    "context": Object {
      "store": Object {
        "lookUpAuthor": [Function],
      },
    },
  },
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {
        "store": Object {
          "lookUpAuthor": [Function],
        },
      },
      "_currentElement": <Article
        article={
                Object {
                        "body": "testBody",
                        "date": "Mon Jul 11 2016 00:00:00 GMT+0000 (UTC)",
                        "title": "testTitle",
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {
          "store": Object {
            "lookUpAuthor": [Function],
          },
        },
        "props": Object {
          "article": Object {
            "body": "testBody",
            "date": "Mon Jul 11 2016 00:00:00 GMT+0000 (UTC)",
            "title": "testTitle",
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <glamorous(div)>
          <glamorous(div)>
                    testTitle
          </glamorous(div)>
          <glamorous(div)>
                    Sun Jul 10 2016
          </glamorous(div)>
          <glamorous(div)>
                    <a
                              href="testWebsite"
                              target="_blank"
                    >
                              testFirstName
                               
                              testLastName
                    </a>
          </glamorous(div)>
          <glamorous(div)>
                    testBody
          </glamorous(div)>
</glamorous(div)>,
        "_debugID": 2,
        "_renderedOutput": <glamorous(div)>
          <glamorous(div)>
                    testTitle
          </glamorous(div)>
          <glamorous(div)>
                    Sun Jul 10 2016
          </glamorous(div)>
          <glamorous(div)>
                    <a
                              href="testWebsite"
                              target="_blank"
                    >
                              testFirstName
                               
                              testLastName
                    </a>
          </glamorous(div)>
          <glamorous(div)>
                    testBody
          </glamorous(div)>
</glamorous(div)>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Article
    article={
        Object {
            "body": "testBody",
            "date": "Mon Jul 11 2016 00:00:00 GMT+0000 (UTC)",
            "title": "testTitle",
          }
    }
/>,
}
`;
